extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/logical/stages/SegmentMapStage.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.msq.logical.stages;
+
+import org.apache.druid.msq.exec.StageProcessor;
+import org.apache.druid.msq.logical.LogicalInputSpec;
+import org.apache.druid.msq.logical.StageMaker;
+import org.apache.druid.sql.calcite.planner.querygen.DruidQueryGenerator.DruidNodeStack;
+import org.apache.druid.sql.calcite.planner.querygen.SourceDescProducer.SourceDesc;
+
+import java.util.List;
+
+public class SegmentMapStage extends AbstractFrameProcessorStage
```

Comment:
Does this existing mean we are creating a stage specifically for applying segment map fns? It seems inefficient, because they could be folded into the next stage. Are you thinking the folding is going to be done with some future work?

Also, does this stage shuffle? I am not quite seeing where that gets set up.

Reply:
these stages doesn't necessarily need to match the physical stages - they do right now; but the `StageMaker` part may optimize it to execute better;
caring about those things at this point wouldn't be easy - but in the `StageMaker` the problem becomes to just identify stages which could be joined....

Reply:
From what you're saying, I take it the current state of this PR is a stepping-stone on the way towards fusing operations together than could be run in the same stage.